drivers/volume/portworx/portworx.go

```diff
@@ -1778,6 +1781,287 @@ func (d *portworx) ValidatePureVolumesNoReplicaSets(volumeName string, params ma
+		}
+		// Ensure there are no WWIDs with empty single paths, as that is a malformed device
+		if len(lsblkEntry.SinglePaths) == 0 {
+			return nil, fmt.Errorf("found mapper %s with no single paths on node %s, this should not happen (we didn't clean something up right?)", mapper, n.MgmtIp)
+		}
+	}
+
+	return lsblkParsed, nil
+}
+
+func (d *portworx) getCurrentPureLocalVolumePaths() (map[string]map[string]pureLocalPathEntry, error) {
+	// Iterate through all nodes, and for each node collect local paths
+	currentDevices := make(map[string]map[string]pureLocalPathEntry)
+	nodes := node.GetStorageDriverNodes()
+	for _, n := range nodes {
+		localPathMap, err := d.collectLocalNodeInfo(n)
+		if err != nil {
+			return nil, err
+		}
+		currentDevices[n.MgmtIp] = localPathMap
+	}
+
+	return currentDevices, nil
+}
+
+// InitializePureLocalVolumePaths sets the baseline for how many Pure devices are attached to each node, so that we can compare against it later
+func (d *portworx) InitializePureLocalVolumePaths() error {
+	currentDevices, err := d.getCurrentPureLocalVolumePaths()
+	if err != nil {
+		return err
+	}
+	logrus.Infof("Pure device baseline initialized to: %+v", currentDevices)
+	d.pureDeviceBaseline = currentDevices
+
+	// TODO: we should only include non-FACD drives here. Eventually, we should handle FACD as its own kind of volume in ValidatePureLocalVolumePaths and validate it matches the pxctl cd l output
+
+	return nil
+}
+
+// ValidatePureLocalVolumePaths checks that the given volumes all have the proper local paths present, *and that no other unexpected ones are present*
+func (d *portworx) ValidatePureLocalVolumePaths() error {
+	t := func() error {
+		currentDevices, err := d.getCurrentPureLocalVolumePaths()
+		if err != nil {
+			return err
+		}
+		logrus.Infof("Current pure devices: %+v", currentDevices)
+
+		// TODO: handle FACD drive failovers properly (these will show up as new devices but they're actually fine)
+		// Remove all devices that are in the baseline (complex set math time!). Also warn if any baseline devices are now missing?
+		for node, baselineDevices := range d.pureDeviceBaseline {
+			currentNodeDevices := currentDevices[node]
+			for wwid := range baselineDevices {
+				if _, exists := currentNodeDevices[wwid]; exists {
+					delete(currentNodeDevices, wwid)
+				} else {
+					return fmt.Errorf("baseline FA device %s originally present on node %s is missing, this should not happen", wwid, node)
+				}
+			}
+		}
+
+		allVolNames, err := d.ListAllVolumes()
+		if err != nil {
+			return err
+		}
+		// Inspect all volumes provided to get the PX spec
+		fadaVolumes := []*api.Volume{}
+		for _, volName := range allVolNames {
+			v, err := d.InspectVolume(volName)
+			if err != nil {
+				return err
+			}
+			if !v.Spec.IsPureVolume() {
+				continue
+			}
+			if v.Spec.ProxySpec.PureBlockSpec == nil {
+				// TODO: handle FBDA volumes properly as well
+				continue
+			}
```

Comment:
No need of this since you are checking for IsPureVolume() ?

Reply:
IsPureVolume() returns true for both FA/FB.

Reply:
What Phani said, yeah, I want FADA specifically